File: LeetCode/Sliding Window/159. Longest Substring with At Most Two Distinct Characters/Solution.dart
import 'dart:math';

class Solution {
  int lengthOfLongestSubstringTwoDistinct(String s) {
      var answer = 0, n = s.length;
      var left = 0, right = 0;
      Map<String, int> added = {};

      while (right < n) {
          added[s[right]] = right;
          if (added.length >= 3){
            var minVal = added.entries.reduce((cur,next) => cur.value < next.value ? cur : next);
            left = minVal.value;
            added.remove(minVal.key);
          }
          answer = max(answer, right - left + 1);
          right += 1;
      }
      
      return answer;
  }
}

File: LeetCode/Sliding Window/159. Longest Substring with At Most Two Distinct Characters/Solution.py
from collections import defaultdict
class Solution:
    def lengthOfLongestSubstringTwoDistinct(self, s: 'str') -> 'int':
        n = len(s)
        if n < 3:
            return n

        answer = 0
        left, right = 0, 0
        added = defaultdict(int)

        while right < n:
            added[s[right]] = right
            if len(added) >= 3:
                val, idx = min(added.items(), key=lambda x: x[1])
                del added[val]
                left = idx + 1

            answer = max(answer, right - left + 1)
            right += 1


        return answer
                

